platform/linux/mbedtls/network_mbedtls_wrapper.c

```diff
@@ -277,6 +279,10 @@ IoT_Error_t iot_tls_connect(Network *pNetwork, TLSConnectParams *params) {
 
 	mbedtls_ssl_conf_read_timeout(&(tlsDataParams->conf), IOT_SSL_READ_TIMEOUT);
 
+#ifdef AWS_IOT_MQTT_SOCKET_NON_BLOCKING
+	mbedtls_net_set_nonblock(&(tlsDataParams->server_fd));
```

Comment:
Does this call affect the underlying configuration of the `send` called by mbedtls (i.e. by calling `send` with non-blocking flag)? 

What is the goal of making it configurable instead of the default behavior of the mbedTLS based implementation of network abstraction?

Reply:
Non-blocking behavior is not per API (like calling send with a flag) - it is set at socket level. Once a socket is marked non-blocking, both send and receive are non-blocking. Here is the implementation of `mbedtls_net_set_nonblock`:

```c
int mbedtls_net_set_nonblock( mbedtls_net_context *ctx )
{
#if ( defined(_WIN32) || defined(_WIN32_WCE) ) && !defined(EFIX64) && \
    !defined(EFI32)
    u_long n = 1;
    return( ioctlsocket( ctx->fd, FIONBIO, &n ) );
#else
    return( fcntl( ctx->fd, F_SETFL, fcntl( ctx->fd, F_GETFL ) | O_NONBLOCK ) );
#endif
}
```

As it is clear from above, the corresponding socket fd is marked as non-blocking and as a result, both send and receive will be non-blocking.

The other is for backward compatibility - Non-blocking behavior is a significant change and I do not want to change the behavior for everyone. Anyone who needs it will opt for it.

Reply:
is there equivalent of  mbedtls_ssl_conf_read_timeout for write operation ? ( mbedtls_ssl_conf_write_timeout ?)  That should make it 'non-blocking' , that is, it should wait for IOT_SSL_WRITE_TIMEOUT and then return error. Having it purely non-blocking may just return immediately and may see the problem that was seen by @aggarw13  ( return success even if the network is down )